### PR TITLE
New dictation modernazation

### PR DESCRIPTION
**Pull request by juggernautsei**
Looking for input on what should I do next.

### CODE REVIEW

File: interface/documents/upload.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+* Drag and Drop file uploader.
+*
+* Copyright (C) 2017 Sherwin Gaddis sherwingaddis@gmail.com
```
Comment (robertdown) [inline interface/documents/upload.php]: We are trying to standardize our docblocks. Examples here http://www.open-emr.org/wiki/index.php/How_to_Document_Your_Code_Properly#Examples_for_OpenEMR

File: interface/documents/upload.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+*
+* @package   OpenEMR
+* @author    Sherwin Gaddis sherwingaddis@gmail.com
+*/
+
+$patient_id = filter_input(INPUT_GET, 'patient_id');
+$category_id = filter_input(INPUT_GET, 'parent_id');
+
+
+$sanitize_all_escapes=true;
```
Comment (robertdown) [inline interface/documents/upload.php]: This functionality has been removed, remove 25, 26

File: interface/documents/upload.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+$fake_register_globals=false;
+
+require_once("../globals.php");
+require_once(dirname(__FILE__) . "/../../library/documents.php");
```
Comment (robertdown) [inline interface/documents/upload.php]: I think '$GLOBALS["srcdir"]' would point to the library folder

File: interface/documents/upload.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    $size     = $_FILES['file']['size'];
+    $owner    = $GLOBALS['userauthorized'];
+
+
+addNewDocument($name,$type,$tmp_name,$error,$size,$owner,$patient_id_or_simple_directory=$patient_id,$category_id,
+	                      $higher_level_path='',$path_depth='1');
+
+}
+
+?>
```
Comment (robertdown) [inline interface/documents/upload.php]: No need for closing tag here

File: interface/forms/dictation/new.php [hunk 5]
```diff
@@ -1,29 +1,56 @@
 <!-- Form generated from formsWiz -->
 <?php
+/** Copyright (C) 2016 Sherwin Gaddis <sherwingaddis@gmail.com>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: See above comment about docblocks

File: interface/forms/dictation/new.php [hunk 6]
```diff
@@ -1,29 +1,56 @@
...
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+<?php Header::setupHeader(); ?>
+	<script type="text/javascript" src="<?php echo $webroot."/library/custom_template/ckeditor/ckeditor.js" ?>"</script>
+	<script src="<?php echo $webroot."/library/custom_template/ckeditor/_samples/sample.js" ?>" type="text/javascript"></script>
+	<link href="<?php echo $webroot."/library/custom_template/ckeditor/_samples.css"; ?>" rel="stylesheet" type="text/css" />
 </head>
 <body class="body_top">
+<h1><?php echo xlt('Speech Dictation'); ?></h1><br><br>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Wrap this in a div with the class page-header

File: interface/forms/dictation/new.php [hunk 7]
```diff
@@ -1,29 +1,56 @@
...
+
+<div class="container">
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Container should be up by the body, we want most everything contained

File: interface/forms/dictation/new.php [hunk 8]
```diff
@@ -1,29 +1,56 @@
...
-<br>
+<div class="form-group">
+
+<label class="text"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" cols=80 rows=24 wrap="virtual" name="dictation" ></textarea>
+</div>
+<div class="form-group">
+<label class="text"><?php echo xlt('Additional Notes:'); ?> </label><br><textarea class="form-control ckeditor" cols=80 rows=8 wrap="virtual" name="additional_notes" ></textarea>
+</div>
+
+<a href="javascript:top.restoreSession();document.my_form.submit();" class="link_submit"><button><?php echo xlt('Save'); ?></button></a>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Remove the a tag, setup the button with a type=submit and the following classes: btn btn-default btn-save

File: interface/forms/dictation/new.php [hunk 9]
```diff
@@ -1,29 +1,56 @@
...
+
 <a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="link"
- onclick="top.restoreSession()">[<?php echo xlt('Don\'t Save'); ?>]</a>
+ onclick="top.restoreSession()"><button><?php echo xlt('Don\'t Save'); ?></button></a>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: This will seem odd, but there's reasons. This element will take the user away from the page so it should be an <a> tag, but there is no need for a button inside. Add the btn btn-cancel and btn-link classes to the anchor. This will make it *look* like a button while maintaining the proper functionality. Also, I'd stick with "Cancel" as the text, that seems to be the paradigm throughout the rest of the program

File: interface/forms/dictation/new.php [hunk 10]
```diff
@@ -1,29 +1,56 @@
...
+<?php Header::setupHeader(); ?>
+	<script type="text/javascript" src="<?php echo $webroot."/library/custom_template/ckeditor/ckeditor.js" ?>"</script>
+	<script src="<?php echo $webroot."/library/custom_template/ckeditor/_samples/sample.js" ?>" type="text/javascript"></script>
+	<link href="<?php echo $webroot."/library/custom_template/ckeditor/_samples.css"; ?>" rel="stylesheet" type="text/css" />
 </head>
 <body class="body_top">
+<h1><?php echo xlt('Speech Dictation'); ?></h1><br><br>
+
+<div class="container">
 <form method=post action="<?php echo $rootdir;?>/forms/dictation/save.php?mode=new" name="my_form">
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Take a look at the examples on http://getbootstrap.com/css/#forms for how to layout a form.

**Comment by robertdown** [comment]
As a note, buttons in bootstrap get their styling via the btn class, not via the button element. it's common to use btn classes on anchor tags, especially cancel links.

**Comment by bradymiller** [comment]
delete the 2 unrelated files to make this PR more clear.

File: interface/forms/dictation/new.php [hunk 11]
```diff
@@ -1,29 +1,56 @@
...
 ?>
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+<?php Header::setupHeader(); ?>
+	<script type="text/javascript" src="<?php echo $webroot."/library/custom_template/ckeditor/ckeditor.js" ?>"</script>
+	<script src="<?php echo $webroot."/library/custom_template/ckeditor/_samples/sample.js" ?>" type="text/javascript"></script>
+	<link href="<?php echo $webroot."/library/custom_template/ckeditor/_samples.css"; ?>" rel="stylesheet" type="text/css" />
```
Comment (bradymiller) [inline interface/forms/dictation/new.php]: Do these things drastically change form? And also, will the form still be backward compatible?

File: interface/documents/upload.php (hunk 3 above)
Comment (juggernautsei) [inline interface/documents/upload.php]: Actually $scrdir will do the job.

File: interface/forms/dictation/new.php (hunk 11 above)
Comment (juggernautsei) [inline interface/forms/dictation/new.php]: To my limited knowledge, the ckeditor wraps around the text area like styling. I will test to make sure but it does not change the property of the text area.

Comment (juggernautsei) [inline interface/forms/dictation/new.php]: Which two unrelated files

Comment (bradymiller) [inline interface/forms/dictation/new.php]: 2 unrelated files are:
interface/documents/upload.php
templates/documents/general_upload.html (revert this to file in master by just copying it over)

File: interface/forms/dictation/report.php [hunk 12]
```diff
@@ -18,8 +39,8 @@ function dictation_report( $pid, $encounter, $cols, $id) {
         $value = "yes";
       }
       $key=ucwords(str_replace("_"," ",$key));
-      print "<td><span class=bold>" . xlt($key) . ": </span><span class=text>" .
-        nl2br(text($value)) . "</span></td>";
+      print "<td><span class='bold'>" . xlt($key) . ": </span><span class='text'>" .
+        nl2br($value) . "</span></td>";
```
Comment (bradymiller) [inline interface/forms/dictation/report.php]: removing the text() call is an issue here since it opens the door to cross scripting attacks.

Comment (juggernautsei) [inline interface/forms/dictation/report.php]: But leaving it there causes the output to show the html characters.  I have changed the order and it still outputs the line breaks to the screen.

File: interface/forms/dictation/report.php [hunk 13]
```diff
@@ -18,8 +39,8 @@ function dictation_report( $pid, $encounter, $cols, $id) {
         $value = "yes";
       }
       $key=ucwords(str_replace("_"," ",$key));
-      print "<td><span class=bold>" . xlt($key) . ": </span><span class=text>" .
-        nl2br(text($value)) . "</span></td>";
+      print "<td><span class='bold'>" . xlt($key) . ": </span><span class='text'>" .
+        html_entity_decode($value, ENT_QUOTES | ENT_HTML5) . "</span></td>";
```
Comment (bradymiller) [inline interface/forms/dictation/report.php]: html_entity_decode won't protect from cross scripting attacks.

Can't have raw html (which then allows raw javascript) in fields like this. This could be a potential fatal flaw of ckeditor if that is what requires bypassing html escaping.

Comment (juggernautsei) [inline interface/forms/dictation/report.php]: I was doing some reading and it is not the job of the editor but the server side store process. So, reading that, I changed the way the data is stored. I put the input filters in place and then converted any HTML into entities and that is stored in the tables. Now we should be safe from XSS attacks.

Comment (bradymiller) [inline interface/forms/dictation/report.php]: This still doesn't sound like it will protect from xss. Let me look into this a bit.

**Comment by juggernautsei** [comment]
xss
I entered a script tag in the text box and it comes back to the screen as text.

**Comment by bradymiller** [comment]
@juggernautsei ,
This is odd. Here's what is happening in the code for the text entries.
1. Note the filter_input() calls are doing no escaping/filtering/sanitizing
2. The htmlspecialchars() is escaping the html elements (like < and >)
3. The html_entity_decode() should then unescape all the html elements.
(thus the above really shouldn't do anything; ie. 1 big circle :) )

I am guessing that ckeditor has a some basic code built in to escape obvious things like \<script\> tags.

Would test it on the prior code without all the above escaping stuff.

**Comment by juggernautsei** [comment]
I put the report file back to the way it was with the text($value) and the save.php is back to the original. Below is what shows in the patient chart.
xss-report

**Comment by bradymiller** [comment]
@juggernautsei ,
That is interesting. Just drop the text() call then, and we should be good(ie. don't worry, this wasn't a big exercise in futility since we learned a lot along the way :) ). It looks like ckeditor is escaping the script tags behind the scenes before placing it in the database. Since we are relying on ckeditor to do the html escaping, will be good idea to get most recent version of it in the codebase. What's the best editor out there (this would be a good time to decide on ideal editor to use).

**Comment by bradymiller** [comment]
@juggernautsei ,
Looks like I can bring in most recent version of ckeditor easily via bower. Just let me know if you want it, and then I'll bring it in.

**Comment by juggernautsei** [comment]
We still have a problem. When I remove the text the cross-scripting works. And it does not show on the screen so it is not stopping or rendering harmless the XSS injection. 
Also, in all my searches for editors ckeditor is always on the list and the current version is 4.7

**SOLUTION**
htmlpurifier.org
I've tried it and it works.  Posting code to show how.

**Comment by bradymiller** [comment]
The htmlpurifier looks promising. I can bring that in via composer (so it's then autoloaded); just let me know and I'll bring it into the codebase. -brady

**Comment by bradymiller** [comment]
Another place we can also use this is in pnotes.

**Comment by juggernautsei** [comment]
Yes, pull the HTML purifier into the code base. It has a symphony plugin.

**Comment by juggernautsei** [comment]
Let me know when the HTML purifier and CKEditor 4.7 gets pulled into the code base. Then, we can revisit this and adjust to the code to the new location of the resources.

**Comment by bradymiller** [comment]
Hi @juggernautsei ,

I just added both packages to the codebase(check out the most recent 2 commit in the main codebase). Note I installed the standard-all package for ckeditor (if needed, can install one of the other packages such as full, if needed in the future).

-brady

File: interface/forms/dictation/view.php [hunk 14]
```diff
@@ -12,6 +12,8 @@
 
 use OpenEMR\Core\Header;
 include_once("../../globals.php");
+require_once("../../../public/assets/htmlpurifier-4.9.2/library/HTMLPurifier.auto.php");
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Did this come in via bower or composer?

File: interface/forms/dictation/view.php [hunk 15]
```diff
@@ -26,18 +28,26 @@
...
+$id = filter_input(INPUT_GET, "id", FILTER_VALIDATE_INT);
 $obj = formFetch("form_dictation", $id);
 ?>
    <form method=post action="<?php echo $rootdir?>/forms/dictation/save.php?mode=update&id=<?php echo attr($id);?>" name="my_form">
      <div class="page-header">
       <h1><?php echo xlt('Speech Dictation Review/Edit'); ?></h1>
      </div>
      <div class="form-group">
-       <label for="dictation"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" cols=80 rows=24 wrap="virtual" name="dictation" ><?php echo html_entity_decode(text($obj{"dictation"}));?></textarea>
+       <label for="dictation"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" cols=80 rows=24 wrap="virtual" name="dictation" ><?php
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Suggest breaking this super long line up. Textarea with "form-control" shouldn't have the colts attribute, CSS takes care of that

File: interface/forms/dictation/view.php [hunk 16]
```diff
@@ -26,18 +28,26 @@
...
-       <label for="dictation"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" cols=80 rows=24 wrap="virtual" name="dictation" ><?php echo html_entity_decode(text($obj{"dictation"}));?></textarea>
+       <label for="dictation"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" cols=80 rows=24 wrap="virtual" name="dictation" ><?php 
+              $config = HTMLPurifier_Config::createDefault();
+              $purifier = new HTMLPurifier($config);
+              $clean_html = $purifier->purify($obj{"dictation"});
+              echo trim($clean_html); ?></textarea>
      </div>
      <div class="form-group">
-       <label for="additional_notes"><?php echo xlt('Additional Notes: '); ?></span><br><textarea class="form-control ckeditor"cols=80 rows=8 wrap="virtual" name="additional_notes" ><?php echo html_entity_decode(text($obj{"additional_notes"}));?></textarea>
+       <label for="additional_notes"><?php echo xlt('Additional Notes: '); ?></span><br><textarea class="form-control ckeditor"cols=80 rows=8 wrap="virtual" name="additional_notes" ><?php
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: See above comment

File: interface/forms/dictation/view.php (hunk 14 above)
Comment (bradymiller) [inline interface/forms/dictation/view.php]: This was brought in via composer in the current codebase

Comment (bradymiller) [inline interface/forms/dictation/view.php]: so, is autoloaded

Comment (juggernautsei) [inline interface/forms/dictation/view.php]: So the ../../../public should be replaced with which global? I looked for it and couldn't find it or remember what it is.

Comment (bradymiller) [inline interface/forms/dictation/view.php]: Since it's autoloaded, no include/require is needed. The class is magically available in all scripts. However, will need to rebase your code on the most current codebase (in order to bring this code in). Also should delete entire public/assets/htmlpurifier-4.9.2 directory from your branch.

File: interface/forms/dictation/report.php [hunk 17]
```diff
@@ -14,9 +14,6 @@
 include_once($GLOBALS["srcdir"]."/api.inc");
 require_once("../../../public/assets/htmlpurifier-4.9.2/library/HTMLPurifier.auto.php");
```
Comment (bradymiller) [inline interface/forms/dictation/report.php]: remove above line

File: interface/documents/upload.php [hunk 18]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (bradymiller) [inline interface/documents/upload.php]: remove this file

File: interface/forms/dictation/new.php [hunk 19]
```diff
@@ -1,29 +1,50 @@
...
 ?>
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+   <?php Header::setupHeader(); ?>
+   <script type="text/javascript" src="<?php echo $webroot."/library/custom_template/ckeditor/ckeditor.js" ?>"</script>
+   <script src="<?php echo $webroot."/library/custom_template/ckeditor/_samples/sample.js" ?>" type="text/javascript"></script>
+   <link href="<?php echo $webroot."/library/custom_template/ckeditor/_samples.css"; ?>" rel="stylesheet" type="text/css" />
```
Comment (bradymiller) [inline interface/forms/dictation/new.php]: after you rebase you code the most recent codebase, then you will use the ckeditor that is in public/assets/

File: interface/forms/dictation/report.php [hunk 20]
```diff
@@ -1,7 +1,29 @@
...
+ * @link    http://www.open-emr.org
+ * @author  Sherwin Gaddis <sherwingaddis@gmail.com>
+ * @license https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ * @copyright Copyright (c) 2017 Sherwin Gaddis <sherwingaddis@gmail.com>
+ * 
+ */
+
 include_once(dirname(__FILE__).'/../../globals.php');
 include_once($GLOBALS["srcdir"]."/api.inc");
+require_once("../../../public/assets/htmlpurifier-4.9.2/library/HTMLPurifier.auto.php");
```
Comment (bradymiller) [inline interface/forms/dictation/report.php]: After you rebase code to most recent codebase, then you can remove the above line.

File: interface/forms/dictation/view.php [hunk 21]
```diff
@@ -1,31 +1,61 @@
...
+ * @copyright Copyright (c) 2017 Sherwin Gaddis <sherwingaddis@gmail.com>
+ * 
+ */
 
-
-
-
+use OpenEMR\Core\Header;
 include_once("../../globals.php");
+require_once("../../../public/assets/htmlpurifier-4.9.2/library/HTMLPurifier.auto.php");
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: After you rebase code to most recent codebase, then you can remove the above line.

File: interface/forms/dictation/view.php [hunk 22]
```diff
@@ -1,31 +1,61 @@
...
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+<title>Review</title>
+   <?php Header::setupHeader(); ?>
+   <script type="text/javascript" src="<?php echo $webroot."/library/custom_template/ckeditor/ckeditor.js" ?>"</script>
+   <script src="<?php echo $webroot."/library/custom_template/ckeditor/_samples/sample.js" ?>" type="text/javascript"></script>
+   <link href="<?php echo $webroot."/library/custom_template/ckeditor/_samples.css"; ?>" rel="stylesheet" type="text/css" />
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: After you rebase code to most recent codebase, then you can use the ckeditor in public/assets/

File: interface/forms/dictation/view.php (hunk 14 above)
Comment (robertdown) [inline interface/forms/dictation/view.php]: Will likely need a `use`statement

Comment (juggernautsei) [inline interface/forms/dictation/view.php]: The use statement was not needed as @bradymiller suggested. It works without the use statement. I will post in a moment.

Comment (robertdown) [inline interface/forms/dictation/view.php]: It shouldn't work without that line. It needs to be added

Comment (juggernautsei) [inline interface/forms/dictation/view.php]: Now I am not sure which line you are referring. If you are referring to this line:
require_once("/public/assets/htmlpurifier-4.9.2/library/HTMLPurifier.auto.php");
I am pretty sure it works without it because without that line the page loads. Please correct me if I am wrong but if the page makes a call to a class that it can't find that page should throw an error and not load, right?  Everything is working fine without that line. 
You can try it out at http://omp.openmedpractice.com/rolebase log in admin pass.

File: interface/forms/dictation/new.php [hunk 23]
```diff
@@ -1,29 +1,51 @@
...
 include_once("$srcdir/api.inc");
 formHeader("Form: dictation");
 $returnurl = 'encounter_top.php';
 ?>
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+  <title>Dictation</title>
```
Comment (bradymiller) [inline interface/forms/dictation/new.php]: translate:
``````
<title><?php echo xlt('Dictation'); ?></title>

File: interface/forms/dictation/new.php [hunk 24]
```diff
@@ -1,29 +1,51 @@
...
+   <?php Header::setupHeader(); ?>
+   <script type="text/javascript" src="<?php echo $webroot."/public/assets/ckeditor-4-7-0/ckeditor.js" ?>"</script>
+   <script src="<?php echo $webroot."/public/assets/ckeditor-4-7-0/js/sample.js" ?>" type="text/javascript"></script>
+   <link href="<?php echo $webroot."/public/assets/ckeditor-4-7-0/css/samples.css"; ?>" rel="stylesheet" type="text/css" />
```
Comment (bradymiller) [inline interface/forms/dictation/new.php]: For the above 3 lines, need to use the global for the assets path(this will allow us to migrate that path easily if needed):
``````
<script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/public/assets/ckeditor-4-7-0/ckeditor.js"</script>
<script src="<?php echo $GLOBALS['assets_static_relative']; ?>/public/assets/ckeditor-4-7-0/js/sample.js"  type="text/javascript"></script>
<link href="<?php echo $GLOBALS['assets_static_relative']; ?>/public/assets/ckeditor-4-7-0/css/samples.css"  rel="stylesheet" type="text/css" />

Comment (bradymiller) [inline interface/forms/dictation/new.php]: note I made some edits to above post

File: interface/forms/dictation/new.php [hunk 25]
```diff
@@ -1,29 +1,51 @@
...
-<br>
-<a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="link"
- onclick="top.restoreSession()">[<?php echo xlt('Don\'t Save'); ?>]</a>
-</form>
+<div class="container">
+   <div class="page-header">
+      <h1><?php echo xlt('Speech Dictation'); ?></h1><br><br>
+   </div>
+
+  <form method=post action="<?php echo $rootdir;?>/forms/dictation/save.php?mode=new" name="my_form">
```
Comment (bradymiller) [inline interface/forms/dictation/new.php]: forms needs to deal with session leak issue (especially since it was removed from the Save link):
``````
<form method=post action="<?php echo $rootdir;?>/forms/dictation/save.php?mode=new" name="my_form" onsubmit="return top.restoreSession()">
``````
More information on this can be found here:
http://www.open-emr.org/wiki/index.php/OpenEMR_System_Architecture#PHP_Sessions_and_Browser_Windows

File: interface/forms/dictation/view.php [hunk 26]
```diff
@@ -1,31 +1,60 @@
...
 include_once("../../globals.php");
+
 $returnurl = 'encounter_top.php';
 ?>
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+<title>Review</title>
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: translate and use Dictation:
``````
<title><?php echo xlt('Dictation'); ?></title>

File: interface/forms/dictation/view.php [hunk 27]
```diff
@@ -1,31 +1,60 @@
...
+   <?php Header::setupHeader(); ?>
+   <script type="text/javascript" src="<?php echo $webroot."/public/assets/ckeditor-4-7-0/ckeditor.js" ?>"</script>
+   <script src="<?php echo $webroot."/public/assets/ckeditor-4-7-0/js/sample.js" ?>" type="text/javascript"></script>
+   <link href="<?php echo $webroot."/public/assets/ckeditor-4-7-0/css/samples.css"; ?>" rel="stylesheet" type="text/css" />
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: For the above 3 lines, need to use the global for the assets path(this will allow us to migrate that path easily if needed):
``````
<script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/ckeditor-4-7-0/ckeditor.js"</script>
<script src="<?php echo $GLOBALS['assets_static_relative']; ?>/ckeditor-4-7-0/js/sample.js" type="text/javascript"></script>
<link href="<?php echo $GLOBALS['assets_static_relative']; ?>/ckeditor-4-7-0/css/samples.css" rel="stylesheet" type="text/css" />

File: interface/forms/dictation/view.php [hunk 28]
```diff
@@ -1,31 +1,60 @@
...
-<span class="title"><?php echo xlt('Speech Dictation'); ?></span><Br><br>
-<span class=text><?php echo xlt('Dictation: '); ?></span><br><textarea cols=80 rows=24 wrap=virtual name="dictation" ><?php echo text($obj{"dictation"});?></textarea><br>
-<span class=text><?php echo xlt('Additional Notes: '); ?></span><br><textarea cols=80 rows=8 wrap=virtual name="additional_notes" ><?php echo text($obj{"additional_notes"});?></textarea><br>
-<br>
-<a href="javascript:top.restoreSession();document.my_form.submit();" class="link_submit">[<?php echo xlt('Save'); ?>]</a>
-<br>
-<a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="link"
- onclick="top.restoreSession()">[<?php echo xlt('Don\'t Save Changes'); ?>]</a>
-</form>
+   <form method=post action="<?php echo $rootdir?>/forms/dictation/save.php?mode=update&id=<?php echo attr($id);?>" name="my_form">
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: forms needs to deal with session leak issue (especially since it was removed from the Save link):
``````
<form method=post action="<?php echo $rootdir?>/forms/dictation/save.php?mode=update&id=<?php echo attr($id);?>" name="my_form" onsubmit="return top.restoreSession()">
``````
More information on this can be found here:
http://www.open-emr.org/wiki/index.php/OpenEMR_System_Architecture#PHP_Sessions_and_Browser_Windows

File: interface/forms/dictation/view.php (hunk 27 above)
Comment (bradymiller) [inline interface/forms/dictation/view.php]: Note that the path to samples.* on your bottom 2 lines is incorrect.

File: interface/forms/dictation/new.php (hunk 24 above)
Comment (bradymiller) [inline interface/forms/dictation/new.php]: Note that the path to samples.* on your bottom 2 lines is incorrect.

File: interface/forms/dictation/view.php [hunk 29]
```diff
@@ -1,31 +1,60 @@
...
+     <div class="page-header">
+      <h1><?php echo xlt('Speech Dictation Review/Edit'); ?></h1>
```
Comment (bradymiller) [inline interface/forms/dictation/view.php]: Probably doesn't makes sense to make this title different on the forms since most forms do not allow this (goal is keeping this uniform), since usually view.php redirects to new.php on a lot of the forms. Also good to keep the title in head the same as this title which should be the same as the form name. So would use 'Speech Dictation' in all those place.

File: interface/forms/dictation/new.php [hunk 30]
```diff
@@ -1,29 +1,50 @@
...
- onclick="top.restoreSession()">[<?php echo xlt('Don\'t Save'); ?>]</a>
-</form>
+<div class="container">
+   <div class="page-header">
+      <h1><?php echo xlt('Speech Dictation'); ?></h1><br><br>
+   </div>
+  <form method=post action="<?php echo $rootdir;?>/forms/dictation/save.php?mode=new" name="my_form" onclick="top.restoreSession()">
+
+     <div class="form-group">
+       <label for="dictation"><?php echo xlt('Dictation: '); ?></label><br><textarea class="form-control ckeditor" name="dictation" ></textarea>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: @bradymiller, correct me if I'm wrong, but shouldn't end-of-line spaces and punctuation like this be avoided? Also, I'm in favor of yanking colons from labels all together

Comment (bradymiller) [inline interface/forms/dictation/new.php]: @robertdown 
In this case, the damage was already done about 8 years ago; meaning, this "constant" was already sent to the translation pipeline in it's current form at that time. So, changing these can cause a new constant that needs to be translated. This isn't always the case; for example, here , the Dictation constant already exists, so changing this wouldn't even create a new constant:
https://github.com/openemr/openemr/blob/master/contrib/util/language_translations/currentConstants.txt#L1899

A good general strategy:
If there's a reason(such as not wanting to have the colons), then should change these constants, however, shouldn't do a codebase wide arbitrary fixing of leading/trailing spacing in contants. Hope that makes some sense.

**Comment by bradymiller** [comment]
Remove the following directory (this will make it much easier to review the code):
public/assets/htmlpurifier-4.9.2

**Comment by juggernautsei** [comment]
Review Please

**Comment by bradymiller** [comment]
This branch wasn't working since missing the ckeditor, so I rebased this code into 1 commit and into the most recent codebase in another PR here:
https://github.com/openemr/openemr/pull/879

Should start working on the new PR branch and will close this PR.
